main/src/library/Vector.flix

```diff
@@ -472,17 +487,81 @@ mod Vector {
 
     ///
     /// Returns the result of applying `f` to every element in `v` along with that element's index.
     ///
     /// That is, the result is of the form: `[ f(a[0], 0), f(a[1], 1), ... ]`.
     ///
+    /// Purity reflective: Runs in parallel when given a pure function `f`
     pub def mapWithIndex(f: (Int32, a) -> b \ ef, v: Vector[a]): Vector[b] \ ef =
+        typematch f {
+            case g: (Int32, a) -> b \ {} =>
+                if (useParallelEvaluation(v))
+                    parMapWithIndex(threads() - 1, g, v)
+                else seqMapWithIndex(g, v)
+            case g: (Int32, a) -> b \ ef => seqMapWithIndex(g, v)
+            case _:_ => unreachable!()
+        }
+
+    ///
+    /// Returns the result of applying `f` to every element in `v`.
+    ///
+    /// The result is a new vector.
+    ///
+    pub def seqMap(f: a -> b \ ef, v: Vector[a]): Vector[b] \ ef =
+        init(i -> f(get(i, v)), length(v))
+
+    ///
+    /// Returns the result of applying `f` to every element in `v`.
+    ///
+    /// The result is a new vector.
+    ///
+    /// The implementation spawns `n` threads each applying `f` on some subvector that is
+    /// disjoint from the rest of the threads.
+    ///
+    pub def parMap(n: Int32, f: a -> b, v: Vector[a]): Vector[b] = parMapWithIndex(n, (_, x) -> f(x), v)
+
+    ///
+    /// Returns the result of applying `f` to every element in `v` along with that element's index.
+    ///
+    /// That is, the result is of the form: `[ f(a[0], 0), f(a[1], 1), ... ]`.
+    ///
+    pub def seqMapWithIndex(f: (Int32, a) -> b \ ef, v: Vector[a]): Vector[b] \ ef =
         init(i -> f(i, get(i, v)), length(v))
 
+    ///
+    /// Returns the result of applying `f` to every element in `v` along with that element's index.
+    ///
+    /// That is, the result is of the form: `[ f(a[0], 0), f(a[1], 1), ... ]`.
+    ///
+    /// The implementation spawns `n` threads each applying `f` on some subvector that is
+    /// disjoint from the rest of the threads.
+    ///
+    pub def parMapWithIndex(n: Int32, f: (Int32, a) -> b, v: Vector[a]): Vector[b] =
+        unchecked_cast(region rc {
+            let result: Array[b, rc] = Array.new(rc, length(v));
+            def transform(left, right) = {
+                if (left <= right) {
+                    Array.put(f(left, get(left, v)), left, result);
+                    transform(left + 1, right)
+                }
+                else ()
+            };
+            def parTransform(t, left, right) =
+                if (t <= 1) transform(left, right)
+                else {
+                    spawn parTransform((t - 2) / 2, left, (left + right) / 2) @ rc;
+                    spawn parTransform((t - 2) / 2, (left + right) / 2, right) @ rc
+                };
+            parTransform(n, 0, length(v));
+            unchecked_cast(result as Vector[b])
+        } as Vector[b] \ {})
```

Comment:
I think what I have in mind is that, if the vector is sufficiently large, then we should start `n` threads (corresponding to the number of CPUs) and then each thread would operate on a segment of the vector in parallel. Probably using `par`:

https://doc.flix.dev/parallelism.html

then each thread would return a new vector corresponding to its block which we would concatenate together.

Reply:
My current approach looks like this
```
pub def parMapWithIndex(n: Int32, f: (Int32, a) -> b, v: Vector[a]): Vector[b] =
    def transform(left, right) = region rc {
        let result = Array.new(rc, right - left);
        def loop(l) =
            if (l < right) {
                Array.put(f(l - left, get(l, v)), l - left, result);
                loop(l + 1)
            }
            else ();
        loop(left);
        Array.toVector(result)
    };
    def parTransform(t, left, right) =
        if (t <= 1) transform(left, right)
        else par (
            v1 <- parTransform((t - 2) / 2, left, (left + right) / 2);
            v2 <- parTransform((t - 2) / 2, (left + right) / 2, right)
        ) yield Vector.append(v1, v2);
    parTransform(n, 0, length(v))
```

Reply:
This is much better, but it still suffers from recursion that constructs a tree.

What we would be doing is creating `n` blocks where if we have `n` threads and the vector has length `m` then each block is `m/n` long (where possibly the last block is shorter than the rest). This way we only end up with very few calls to `Vector.append`. Does that make sense?

Reply:
Yes, this makes sense. I'll make this change